### PR TITLE
Fix division by zero NaN in categorical_crossentropy

### LINKED ISSUE #23551: Division by zero NaN vulnerability in categorical_crossentropy
### Description of the Vulnerability

There is a division-by-zero NaN vulnerability in the `categorical_crossentropy` loss function across all Keras 3 backends (Legacy, TensorFlow, PyTorch, JAX, NumPy, OpenVINO). 

When the model predictions (outputs) sum to exactly zero, the normalization step:
`output = output / reduce_sum(output, axis, keepdims=True)`
results in a division by zero. This immediately propagates `NaN`s through the loss and poisons the gradients, causing catastrophic training failure. This edge case is particularly prevalent when models are fine-tuned on resource-constrained hardware or with extreme learning rates where outputs might collapse to zero across the class dimension.

### Proposed Mitigation

This can be seamlessly mitigated by adding `backend.epsilon()` to the denominator prior to division, ensuring numerical stability without impacting the loss calculation for valid distributions:

```python
output = output / (reduce_sum(output, axis, keepdims=True) + backend.epsilon())
```

### Upstream Context

This issue was originally identified and patched in the TensorFlow repository, but the maintainers requested the fix be ported upstream to Keras to protect the wider ecosystem. 

- **Original TensorFlow PR**: https://github.com/tensorflow/tensorflow/pull/123494

I have opened a PR to address this across all Keras backends.

### PR DESCRIPTION
**Pull request by Deeven-Seru**
This PR fixes the zero-day division by zero NaN vulnerability described in #23551. 

By adding `backend.epsilon()` to the denominator when normalizing predictions across all backends (Legacy, TF, Torch, JAX, NumPy, OpenVINO), we prevent catastrophic gradient poisoning when predictions sum to exactly zero.

This issue was originally identified and patched in TensorFlow (https://github.com/tensorflow/tensorflow/pull/123494) and is being ported upstream here per the maintainers' request.

Fixes #23551

## Contributor Agreement
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: test_nan.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# Try to reproduce NaN in categorical_crossentropy
+target = np.array([[1.0, 0.0, 0.0]])
+output = np.array([[0.0, 0.0, 0.0]], dtype=np.float16)
+
+loss = keras.losses.categorical_crossentropy(target, output)
+print("float16:", loss)
+
+output = np.array([[0.0, 0.0, 0.0]], dtype=np.float32)
+loss = keras.losses.categorical_crossentropy(target, output)
+print("float32:", loss)
```
Comment (hertschuh) [inline test_nan.py]: Please create a test function for this within this class:
https://github.com/keras-team/keras/blob/master/keras/src/losses/losses_test.py#L1113

Then, don't do `print` statements, instead verify that there are no NaNs like this: 
```python
        self.assertFalse(np.any(np.isnan(ops.convert_to_numpy(loss))))
```

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for looking into this!

File: test_nan.py (hunk 1 above)
Comment (Deeven-Seru) [reply test_nan.py]: done 👍

**Comment by Deeven-Seru** [comment]
done with requested modification 👍

File: test_nan_jax.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+from keras.src import ops
+
+y_true = np.array([[0, 1, 0], [0, 0, 1]])
+y_pred = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 0.0]])
+loss_fn = losses.CategoricalCrossentropy()
+loss = loss_fn(y_true, y_pred)
+loss_val = ops.convert_to_numpy(loss)
+print("loss_val:", loss_val)
+print("type:", type(loss_val))
+print("isnan.any():", np.isnan(loss_val).any())
```
Comment (hertschuh) [inline test_nan_jax.py]: Remove this file.

**Review by hertschuh — COMMENTED** [review]
The JAX tests are failing, apparently, there are still NaNs in the result. Can you look into this?

**Comment by Deeven-Seru** [comment]
Hi @hertschuh , trying to figure out a fix for this, any suggestions?

**Comment by hertschuh** [comment]
> Hi @hertschuh , trying to figure out a fix for this, any suggestions?

Can you try if `divide_no_nan` fixes it with JAX?

**Comment by Deeven-Seru** [comment]
@hertschuh PTAL :)

**Comment by hertschuh** [comment]
@Deeven-Seru 

> @hertschuh PTAL :)

Oh, I was thinking you would do `divide_no_nan` everywhere for all backends for consistency. Does that not work?

**Comment by Deeven-Seru** [comment]
> @Deeven-Seru
> 
> > @hertschuh PTAL :)
> 
> Oh, I was thinking you would do `divide_no_nan` everywhere for all backends for consistency. Does that not work?

I tried Using divide_no_nan but  in NumPy eagerly evaluates the division before masking, which still throws a RuntimeWarning: divide by zero on execution. Furthermore, PyTorch and TF divide_no_nan implementations allocate multiple intermediate mask tensors, making them slower than a single highly-optimized maximum kernel. So maximum preserves faster continuous gradients for those backends, making divide_no_nan strictly a JAX-specific compiler workaround.

**Comment by hertschuh** [comment]
> > @Deeven-Seru
> > > @hertschuh PTAL :)
> > 
> > 
> > Oh, I was thinking you would do `divide_no_nan` everywhere for all backends for consistency. Does that not work?
> 
> I tried Using divide_no_nan but in NumPy eagerly evaluates the division before masking, which still throws a RuntimeWarning: divide by zero on execution. Furthermore, PyTorch and TF divide_no_nan implementations allocate multiple intermediate mask tensors, making them slower than a single highly-optimized maximum kernel. So maximum preserves faster continuous gradients for those backends, making divide_no_nan strictly a JAX-specific compiler workaround.

Hmm... Ok, looking at the math, I like the epsilon approach better actually. But now I don't like that JAX is inconsistent in the approach.

Sorry for the back and forth, but can you revert back the JAX implementation to the epsilon approach? Let's debug why it's not passing the tests.

**Comment by Deeven-Seru** [comment]
the tests are failing due to float16 underflow I guess , I tried to resolve it by adding custom_vjp to compute in float32

File: keras/src/backend/jax/nn.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
 )
 from keras.src.backend.jax.core import cast
 from keras.src.backend.jax.core import convert_to_tensor
+from keras.src.backend.jax.numpy import divide_no_nan
```
Comment (mathewOracle) [inline keras/src/backend/jax/nn.py:32]: unused import — not referenced anywhere in this diff, grep confirms it

File: keras/src/backend/jax/nn.py [hunk 4]
```diff
@@ -962,6 +963,50 @@ def multi_hot(x, num_classes, axis=-1, dtype=None, sparse=False):
...
+        epsilon_ = convert_to_tensor(backend.epsilon(), dtype=output.dtype)
+        output = output / jnp.maximum(
+            jnp.sum(output, axis, keepdims=True), epsilon_
+        )
+        output = jnp.clip(output, epsilon_, 1.0 - epsilon_)
+        log_prob = jnp.log(output)
+    return -jnp.sum(target * log_prob, axis=axis)
+
+
+categorical_crossentropy_vjp = jax.custom_vjp(
```
Comment (mathewOracle) [inline keras/src/backend/jax/nn.py:981]: this isn't mentioned in the PR description

**Review by mathewOracle — CHANGES_REQUESTED** [review]

File: keras/src/backend/jax/nn.py (hunk 4 above)
Comment (Deeven-Seru) [reply keras/src/backend/jax/nn.py:981]: I will change the description once the fix is confirmed


> this isn't mentioned in the PR description

File: keras/src/backend/jax/nn.py (hunk 3 above)
Comment (Deeven-Seru) [reply keras/src/backend/jax/nn.py:32]: Iteratively working on thr pr

**Comment by Deeven-Seru** [comment]
Hi @hertschuh,

I've updated the PR per your request:
1. **Reverted JAX to the unified epsilon approach**: Removed `custom_vjp` and aligned JAX with the other backends using `jnp.maximum(jnp.sum(output, axis, keepdims=True), epsilon_)`. All distillation and training tests now pass cleanly without any JAX tracer issues.
2. **Fixed OpenVINO constant element type**: Passed `output.get_element_type()` to `ov_opset.constant(backend.epsilon(), ...)` to prevent dtype mismatch errors.

All tests across JAX, NumPy, PyTorch, and TensorFlow pass cleanly. Ready for your review!
